File: api/src/main/java/com/ruby/rubia_server/core/controller/ZApiActivationController.java
package com.ruby.rubia_server.core.controller;

import com.ruby.rubia_server.core.entity.*;
import com.ruby.rubia_server.core.service.ZApiActivationService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/zapi/activation")
@RequiredArgsConstructor
@Slf4j
public class ZApiActivationController {

    private final ZApiActivationService activationService;

    @GetMapping("/status")
    public ResponseEntity<ZApiStatus> getStatus() {
        ZApiStatus status = activationService.getInstanceStatus();
        return ResponseEntity.ok(status);
    }

    @GetMapping("/qr-code/bytes")
    public ResponseEntity<byte[]> getQrCodeBytes() {
        QrCodeResult result = activationService.getQrCodeBytes();
        
        if (result.isSuccess()) {
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.IMAGE_PNG);
            headers.setContentDispositionFormData("attachment", "qrcode.png");
            
            return ResponseEntity.ok()
                .headers(headers)
                .body((byte[]) result.getData());
        } else {
            return ResponseEntity.badRequest().build();
        }
    }

    @GetMapping("/qr-code/image")
    public ResponseEntity<QrCodeResult> getQrCodeImage() {
        QrCodeResult result = activationService.getQrCodeImage();
        return ResponseEntity.ok(result);
    }

    @GetMapping("/phone-code/{phone}")
    public ResponseEntity<PhoneCodeResult> getPhoneCode(@PathVariable String phone) {
        PhoneCodeResult result = activationService.getPhoneCode(phone);
        return ResponseEntity.ok(result);
    }

    @PostMapping("/restart")
    public ResponseEntity<Map<String, Object>> restartInstance() {
        boolean success = activationService.restartInstance();
        
        return ResponseEntity.ok(Map.of(
            "success", success,
            "message", success ? "Instance restarted successfully" : "Failed to restart instance"
        ));
    }

    @PostMapping("/disconnect")
    public ResponseEntity<Map<String, Object>> disconnectInstance() {
        boolean success = activationService.disconnectInstance();
        
        return ResponseEntity.ok(Map.of(
            "success", success,
            "message", success ? "Instance disconnected successfully" : "Failed to disconnect instance"
        ));
    }

    @PostMapping("/webhook/connected")
    public ResponseEntity<String> handleConnectedWebhook(@RequestBody Map<String, Object> payload) {
        try {
            log.info("Z-API instance connected: {}", payload);
            
            String instanceId = (String) payload.get("instanceId");
            Boolean connected = (Boolean) payload.get("connected");
            
            if (Boolean.TRUE.equals(connected)) {
                log.info("Instance {} successfully connected to WhatsApp", instanceId);
            }
            
            return ResponseEntity.ok("OK");
            
        } catch (Exception e) {
            log.error("Error processing connected webhook: {}", e.getMessage(), e);
            return ResponseEntity.badRequest().body("Error: " + e.getMessage());
        }
    }

    @PostMapping("/webhook/disconnected")
    public ResponseEntity<String> handleDisconnectedWebhook(@RequestBody Map<String, Object> payload) {
        try {
            log.info("Z-API instance disconnected: {}", payload);
            
            String instanceId = (String) payload.get("instanceId");
            String error = (String) payload.get("error");
            
            log.warn("Instance {} disconnected. Error: {}", instanceId, error);
            
            return ResponseEntity.ok("OK");
            
        } catch (Exception e) {
            log.error("Error processing disconnected webhook: {}", e.getMessage(), e);
            return ResponseEntity.badRequest().body("Error: " + e.getMessage());
        }
    }
}

File: api-dotnet/Rubia.Server/Controllers/ZApiActivationController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Rubia.Server.Entities;
using Rubia.Server.Services;
using System.Text.Json;

namespace Rubia.Server.Controllers;

[ApiController]
[Route("api/zapi/activation")]
[Authorize]
public class ZApiActivationController : ControllerBase
{
    private readonly ZApiActivationService _activationService;
    private readonly ILogger<ZApiActivationController> _logger;

    public ZApiActivationController(
        ZApiActivationService activationService,
        ILogger<ZApiActivationController> logger)
    {
        _activationService = activationService;
        _logger = logger;
    }

    [HttpGet("status")]
    public async Task<ActionResult<ZApiStatus>> GetStatus()
    {
        var status = await _activationService.GetInstanceStatusAsync();
        return Ok(status);
    }

    [HttpGet("qr-code/bytes")]
    public async Task<IActionResult> GetQrCodeBytes()
    {
        var result = await _activationService.GetQrCodeBytesAsync();

        if (result.Success && result.Data is byte[] bytes)
        {
            Response.Headers.Add("Content-Disposition", "attachment; filename=qrcode.png");
            return File(bytes, "image/png");
        }

        return BadRequest(new { error = result.Error });
    }

    [HttpGet("qr-code/image")]
    public async Task<ActionResult<QrCodeResult>> GetQrCodeImage()
    {
        var result = await _activationService.GetQrCodeImageAsync();
        return Ok(result);
    }

    [HttpGet("phone-code/{phone}")]
    public async Task<ActionResult<PhoneCodeResult>> GetPhoneCode(string phone)
    {
        var result = await _activationService.GetPhoneCodeAsync(phone);
        return Ok(result);
    }

    [HttpPost("restart")]
    public async Task<ActionResult<object>> RestartInstance()
    {
        var success = await _activationService.RestartInstanceAsync();

        return Ok(new
        {
            success,
            message = success ? "Instance restarted successfully" : "Failed to restart instance"
        });
    }

    [HttpPost("disconnect")]
    public async Task<ActionResult<object>> DisconnectInstance()
    {
        var success = await _activationService.DisconnectInstanceAsync();

        return Ok(new
        {
            success,
            message = success ? "Instance disconnected successfully" : "Failed to disconnect instance"
        });
    }

    [HttpPost("webhook/connected")]
    public ActionResult<string> HandleConnectedWebhook([FromBody] Dictionary<string, object> payload)
    {
        try
        {
            _logger.LogInformation("Z-API instance connected: {Payload}", 
                JsonSerializer.Serialize(payload));

            if (payload.TryGetValue("instanceId", out var instanceIdObj) &&
                payload.TryGetValue("connected", out var connectedObj))
            {
                var instanceId = instanceIdObj?.ToString();
                var connected = connectedObj is bool connectedBool && connectedBool;

                if (connected)
                {
                    _logger.LogInformation("Instance {InstanceId} successfully connected to WhatsApp", instanceId);
                }
            }

            return Ok("OK");
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error processing connected webhook: {Message}", e.Message);
            return BadRequest($"Error: {e.Message}");
        }
    }

    [HttpPost("webhook/disconnected")]
    public ActionResult<string> HandleDisconnectedWebhook([FromBody] Dictionary<string, object> payload)
    {
        try
        {
            _logger.LogInformation("Z-API instance disconnected: {Payload}", 
                JsonSerializer.Serialize(payload));

            if (payload.TryGetValue("instanceId", out var instanceIdObj) &&
                payload.TryGetValue("error", out var errorObj))
            {
                var instanceId = instanceIdObj?.ToString();
                var error = errorObj?.ToString();

                _logger.LogWarning("Instance {InstanceId} disconnected. Error: {Error}", instanceId, error);
            }

            return Ok("OK");
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error processing disconnected webhook: {Message}", e.Message);
            return BadRequest($"Error: {e.Message}");
        }
    }
}